File: LeetCode/0206-reverse-linked-list/0206-reverse-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* result;
    void rec(ListNode* prev, ListNode* cur) {
        if (!cur) {
            result = prev;
            return;
        }
        rec(cur, cur->next);
        cur->next = prev;
        prev->next = nullptr;
    }
    ListNode* reverseList(ListNode* head) {
        if (!head)
            return NULL;
        rec(head, head->next);
        return result;
    }
};

File: LeetCode/0206-reverse-linked-list/0206-reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rec(self, prev: Optional[ListNode], head: Optional[ListNode]) -> Optional[ListNode]:
        if head.next == None:
            head.next = prev
            return head

        next_head = head.next
        head.next = prev

        return self.rec(head, next_head)
        
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head == None:
            return head

        next_head = head.next
        head.next = None

        return self.rec(head, next_head)